### PR TITLE
Enhance remove command

### LINKED ISSUE #80: Add ddb remove only in index flag
This flag allows the remove command to skip the file existence check. It will remove the file only from the index.

### PR DESCRIPTION
**Pull request by HeDo88TH**
This fixes #80 still a draft but working. 
You should anyway pass the full path of the file to the remove command, example:

`ddb remove -c -d "C:\Users\IT 2\Pictures\DJI" -p "C:\Users\IT 2\Pictures\DJI\DJI_0339.JPG"`

We need to edit the bindings too, do not merge yet please!

### CODE REVIEW

File: src/ddb.cpp [hunk 1]
```diff
@@ -95,13 +95,13 @@ DDB_C_BEGIN
 DDB_C_END
 }
 
-DDBErr DDBRemove(const char *ddbPath, const char **paths, int numPaths, bool recursive){
+DDBErr DDBRemove(const char *ddbPath, const char **paths, int numPaths, bool recursive, bool cached){
 DDB_C_BEGIN
     auto db = ddb::open(std::string(ddbPath), true);
     std::vector<std::string> pathList(paths, paths + numPaths);
     ddb::removeFromIndex(db.get(), ddb::expandPathList(pathList,
```
Comment (pierotofy) [inline src/ddb.cpp:102]: Careful here; if a user uses the recursive flag, expandPathList will still ping the filesystem.

**Review by pierotofy — COMMENTED** [review]
I'm starting to think that perhaps we should modify the behavior of `ddb remove` to attempt to remove files that are referenced in the index without relying on the FS.

For example:


```
[piero:/data … ton/images/test] $ ddb init
Initialized empty database in .ddb
[piero:/data … ton/images/test] $ ddb add test.txt
A	test.txt
[piero:/data … ton/images/test] $ rm test.txt
[piero:/data … ton/images/test] $ ddb rm test.txt
Path does not exist: test.txt
```

This is a problem that shows an architecture fault.

I think `ddb remove` should:
 - Use the filesystem (as it currently works), if files are present.
 - If files are missing, it should still delete files from the index (fixing the behavior shown above)

This would eliminate the need for an extra flag.

Note there are some interesting cases to handle if the filesystem is missing, namely, recursion of directories (one needs to search the ddb index for child directories).

Thoughts?

**Comment by pierotofy** [comment]
Recursive scenario:

```
[piero:/data … ton/images/test] $ tree .
.
├── a
│   ├── b
│   │   └── test.txt
│   └── test.txt
└── test.txt
```

```
[piero:/data … ton/images/test] $ ddb add -r .
A	a/b/test.txt
A	a/test.txt
A	test.txt
A	a
A	a/b
[piero:/data … ton/images/test] $ rm -r a
[piero:/data … ton/images/test] $ ddb remove -r a
Path does not exist: a
```

(Should remove:)

```
D	a/b/test.txt
D      a/test.txt
D	a
D	a/b
```

**Comment by HeDo88TH** [comment]
I think that the `remove` command should work only on the index without checking the file system. If the user needs to synchronize the index he can use the `sync` command.

**Comment by pierotofy** [comment]
Mm, yes that makes sense. In the end, `rm` only operates on the index and does not need the fs.

**Comment by HeDo88TH** [comment]
> Mm, yes that makes sense. In the end, `rm` only operates on the index and does not need the fs.

So the plan is to remove the "cached" option and make the command list and remove all the matching entries in the database.
Do you want to handle it, or I'll do?

**Comment by pierotofy** [comment]
> So the plan is to remove the "cached" option and make the command list and remove all the matching entries in the database.

Yes.

> Do you want to handle it, or I'll do?

I could use some help :+1:

**Comment by HeDo88TH** [comment]
I reverted the previous changes, If you don't mind implementing it because you are more familiar with C++ and it'll take definitely less time.

**Comment by pierotofy** [comment]
I don't think I'll have time to do this this week.

**Comment by HeDo88TH** [comment]
Since we cannot rely on the filesystem to list files we should find a way to convert a path query (with wildcards and / or recursive) to a SQL DELETE. I was writing the changes and I found the following cases.

These are the files in the index:

```
DJI_0361.JPG
DJI_0362.JPG
DJI_0363.JPG
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
output/test
```

**`ddb remove *.jpg`**

```
-- DELETE FROM Entries WHERE Path LIKE '%.jpg' AND depth = 0
SELECT * FROM Entries WHERE Path LIKE '%.jpg' AND depth = 0
```

```
DJI_0361.JPG
DJI_0362.JPG
DJI_0363.JPG
```

**`ddb remove output`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output%'
SELECT * FROM Entries WHERE Path LIKE 'output%' 
```

```
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output
```

**`ddb remove output/*.jpg`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output/%.jpg' AND Depth = 1
SELECT * FROM Entries WHERE Path LIKE 'output/%.jpg' AND Depth = 1
```

```
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
```

**`ddb remove */test/*.jpg`**

```
-- DELETE FROM Entries WHERE Path LIKE '%/test/%.jpg' 
SELECT * FROM Entries WHERE Path LIKE '%/test/%.jpg'
```

```
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
```

**`ddb remove output/test/*.jpg`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output/test/%.jpg' AND Depth = 2
SELECT * FROM Entries WHERE Path LIKE 'output/test/%.jpg' AND Depth = 2
```

```
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
```

**`ddb remove output/test`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output/test%' AND (Depth = 1 OR Depth = 2)
SELECT * FROM Entries WHERE Path LIKE 'output/test%' AND (Depth = 1 OR Depth = 2)
```

```
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
output/test
```

**`ddb remove *.jpg --recursive`**

```
-- DELETE FROM Entries WHERE Path LIKE '%.jpg'
SELECT * FROM Entries WHERE Path LIKE '%.jpg' 
```

```
DJI_0361.JPG
DJI_0362.JPG
DJI_0363.JPG
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
```

**`ddb remove output/*.jpg --recursive`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output/%.jpg' 
SELECT * FROM Entries WHERE Path LIKE 'output/%.jpg'
```

```
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
```

**`ddb remove output --recursive`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output%'
SELECT * FROM Entries WHERE Path LIKE 'output%'
```

```
output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
output/test
```

**`ddb remove output/test --recursive`**

```
-- DELETE FROM Entries WHERE Path LIKE 'output/test%'
SELECT * FROM Entries WHERE Path LIKE 'output/test%'
```

```
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
output/test

```

@pierotofy can you review these scenarios?

**Comment by pierotofy** [comment]
We have to identify two scenarios for each:

1. Programmatic approach (command invoked from bindings/API)
2. Command line invocation

In this case:

`ddb remove *.jpg`

The root directory **depends** on the directory in which `ddb` is invoked **or** if the `-d,--directory` flag is specified.

To illustrate:

```
cd test && ddb remove *.jpg
```

Is not the same as:

```
cd abc && ddb remove *.jpg
```

The former needs to generate:

```
-- DELETE FROM Entries WHERE Path LIKE 'test/%.jpg' AND depth = 1
```

The latter needs to generate:

```
-- DELETE FROM Entries WHERE Path LIKE 'abc/%.jpg' AND depth = 1
```

Invocation from API bindings should expose a "root" parameter that specifies the root reference directory (or default to "").

**Comment by pierotofy** [comment]
Likewise for `ddb remove output`, or more generally `ddb remove <directory>`, <directory> is relative to `-d` or to the current path where ddb is invoked.

Assuming `ddb remove output` (or `ddb remove output/../output/./../output/`) is invoked in the root directory (same directory as the one containing `.ddb`), then you have two cases:

1. If output is a `directory`;

-- DELETE FROM Entries WHERE Path LIKE 'output%'
SELECT * FROM Entries WHERE Path LIKE 'output%' 

(Is correct)

2. If output is a `folder` (check the `type` enum):

Then the query should be:

```
-- DELETE FROM Entries WHERE Path LIKE 'output%' WHERE depth = 0
SELECT * FROM Entries WHERE Path LIKE 'output%' WHERE depth = 0
```

output/DJI_0339.jpg
output/DJI_0362.jpg
output/DJI_0363.jpg
output
output/test/DJI_0361.jpg
output/test/DJI_0362.jpg
output/test/DJI_0363.jpg
output/test

**Comment by pierotofy** [comment]
> ddb remove output/*.jpg
> 
> -- DELETE FROM Entries WHERE Path LIKE 'output/%.jpg' AND Depth = 1
> SELECT * FROM Entries WHERE Path LIKE 'output/%.jpg' AND Depth = 1
> 
> output/DJI_0339.jpg
> output/DJI_0362.jpg
> output/DJI_0363.jpg

Yes. Also `ddb remove output\*.jpg` (on Windows) should work the same.

**Comment by pierotofy** [comment]
> ddb remove */test/*.jpg
> 
> -- DELETE FROM Entries WHERE Path LIKE '%/test/%.jpg' 
> SELECT * FROM Entries WHERE Path LIKE '%/test/%.jpg'
> 
> output/test/DJI_0361.jpg
> output/test/DJI_0362.jpg
> output/test/DJI_0363.jpg
> 

This gets tricky and I wouldn't even add it as a valid wildcard expression; `*` could refer to the top level directories, or to the top level directories and their subdirectories. Here's a case where this fails:

`output/test/abc/test.jpg`

The wildcard expression above would remove the file, when it probably shouldn't have.

**Comment by pierotofy** [comment]
I'd say we can leave aside the `--recursive` scenarios to simplify things; if a path is a directory, then we recursively delete it, if it's a file we don't (already covered in the cases above). If one wants to nuke the repository, they could invoke:

`ddb remove .` (which is a directory, thus recursive).

**Comment by pierotofy** [comment]
Importantly, one should be able to invoke: `ddb remove *.JPG *.txt` to remove all JPG and txt files in the current directory.

**Comment by pierotofy** [comment]
I'd say, definitely think about edge cases. This is an easy function where bugs could be introduced (big ones, since they involve data deletion)

1. Remember you can't trust file extensions! Files can be with or without extension and directories can have extensions.
2. Files could share the same name as directories (e.g. `test` and `test.tif`)
3. Translating `*` to `%` may work for simple cases, but think about the edge cases.

**Comment by HeDo88TH** [comment]
Thanks for the comments @pierotofy , I wanted to check with you to be on the same page as this behavior is not as trivial as it seems at first glance

**Comment by pierotofy** [comment]
Yep; it's not trivial, path traversal has a lots of quirks at edge cases. Unit testing will help a bunch here. (See `fs_test.cpp` for similar examples related to FS issues).

**Comment by HeDo88TH** [comment]
The implementation is almost complete, it is giving me this error:

```
C:\Users\Art Director\source\repos\DroneDB\build>ddb remove -d "C:\Users\Art Director\Documents\temp" -p "C:\Users\Art Director\Documents\temp\pics"
Root directory: "C:\\Users\\Art Director\\Documents\\temp"
Deleting path: pics
Query: pics
        Deleting Match: pics
D       pics
        Deleting Match: pics/%
database is locked
```

What's using the database? help @pierotofy

**Comment by pierotofy** [comment]
You are probably in the middle of a TRANSACTION while you're doing SELECT and other queries.

A transaction should only be initiated once you start deleting entries.

That's my guess.

File: test/ddb_test.cpp [hunk 2]
```diff
@@ -43,18 +43,306 @@ TEST(getIndexPathList, dontIncludeDirs) {
...
+
+    db.open(dbPath.string());
+
+    std::vector<std::string> toRemove;
+    // 1
+    toRemove.emplace_back((testFolder / "..." / "1JI_0065.JPG").string());
+
+    removeFromIndex(&db, toRemove);
+
+	// Straaaaaange
```
Comment (HeDo88TH) [inline test/ddb_test.cpp:339]: Why is it removing the "..."? It works but it shouldn't!

**Comment by HeDo88TH** [comment]
I added a bunch of tests that all pass, I will add more to cover some edge cases. I just found out one that, IMO, should not work but it does. Do you have any hints? @pierotofy

**Comment by pierotofy** [comment]
`...` is not a valid folder, so the test is actually correct.

I've fixed that particular one.

**Comment by HeDo88TH** [comment]
I think that mixing forward and back slashes is not legal in a path. How can the path figure out the right path separator?

```
    EXPECT_EQ(countEntries(db.get()), 6);
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2-sub"), 1);
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2-sub/test"), 1);

    toRemove.emplace_back((testFolder / "a b\\1" / "test%2" / ".." / "test%2" / "." / "%2-*").string());
    removeFromIndex(db.get(), toRemove);

    // Two entries (1 folder, 1 file) should have been removed
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2-sub"), 0);
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2-sub/test"), 0);

    EXPECT_EQ(countEntries(db.get()), 4);

    // Before
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2"), 1);

    toRemove.clear();
    toRemove.emplace_back((testFolder / "a b\\1" / "test%2" / "%2").string());

    removeFromIndex(db.get(), toRemove);

    // Only one file should have been removed
    EXPECT_EQ(countEntries(db.get(), "a b\\1/test%2/%2"), 0);
    EXPECT_EQ(countEntries(db.get()), 3);

```

I added a sanitizing step for the query param. I'm testing using the "*" as the wildcard but I cannot figure out how to pass it as a parameter due to linux command line parameters expansion.

**Comment by pierotofy** [comment]
```
[piero:/data … e/brighton/test] $ touch 'a\b'
[piero:/data … e/brighton/test] $ ls
'a\b'   test.txt
[piero:/data … e/brighton/test] $ 
```

**Comment by pierotofy** [comment]
```
# ddb info '*.txt'
*.txt does not exist
```

**Comment by pierotofy** [comment]
Call https://en.cppreference.com/w/cpp/experimental/fs/path/generic_string to always get a Unix style path so you don't have to worry about Windows.

**Comment by HeDo88TH** [comment]
The problem is that when a path contains both forward and back slashes the generic() method does not recognise the difference between a '\' that is part of a file name and a '\' that is a windows path separator and then it converts both to '/'. This makes the matching impossible with the database entries that still contain '\'

**Comment by pierotofy** [comment]
Thinking about it, even if we do support "\" characters in filenames, there will be problems syncing to Windows clients since WIndows does not allow files with such characters. 

I've opened https://github.com/DroneDB/DroneDB/issues/94

So for now, assume that the paths in the index do not contain such characters and don't worry about them.

You can adjust/remove my "hardcase" test to reflect that.

**Comment by pierotofy** [comment]
Fun fact, Dropbox will simply rename "a\b" to "b" when you upload a file and try to sync it to Windows (but we don't want to do that).

**Comment by HeDo88TH** [comment]
> Thinking about it, even if we do support "`\`" characters in filenames, there will be problems syncing to Windows clients since WIndows does not allow files with such characters.
> 
> I've opened #94
> 
> So for now, assume that the paths in the index do not contain such characters and don't worry about them.
> 
> You can adjust/remove my "hardcase" test to reflect that.

Ok, I was starting to lose sleep on it.

> Fun fact, Dropbox will simply rename "a\b" to "b" when you upload a file and try to sync it to Windows (but we don't want to do that).

NICE!

**Comment by HeDo88TH** [comment]
The tests keep failing on

```
* Issue another request to this URL: 'http://ot-data1.sdsc.edu:9090/otr/getdem?demtype=AW3D30&west=-92.060910&south=46.798175&east=-91.928101&north=46.886999&outputFormat=GTiff'
*   Trying 198.202.90.171:9090...
* connect to 198.202.90.171 port 9090 failed: Connection refused
* Failed to connect to ot-data1.sdsc.edu port 9090: Connection refused
* Closing connection 3
unknown file: error: C++ exception with description "Couldn't connect to server: Failed to connect to ot-data1.sdsc.edu port 9090: Connection refused" thrown in the test body.
[  FAILED  ] dsmServiceAltitude.Normal (1946 ms)
```

On my PC it is passing. 
I suppose it fails on github because of the custom port (9090)? @pierotofy

**Comment by pierotofy** [comment]
Server must be down: http://ot-data1.sdsc.edu:9090/otr/getdem?demtype=AW3D30&west=-92.060910&south=46.798175&east=-91.928101&north=46.886999&outputFormat=GTiff

**Comment by pierotofy** [comment]
I need to setup a DSM service that is not reliant on third parties...

**Comment by HeDo88TH** [comment]
> I need to setup a DSM service that is not reliant on third parties...

Should I comment that test out by now?
Otherwise is it safe to merge?

**Comment by pierotofy** [comment]
I'm working on a fix for the DSM service.

Can we fix this one? 

```
/home/piero/Documents/ddb/src/entry.h:48:38: warning: ignoring return value of ‘static nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer> nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer>::parse(nlohmann::detail::input_adapter&&, nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer>::parser_callback_t, bool) [with ObjectType = std::map; ArrayType = std::vector; StringType = std::__cxx11::basic_string<char>; BooleanType = bool; NumberIntegerType = long int; NumberUnsignedType = long unsigned int; NumberFloatType = double; AllocatorType = std::allocator; JSONSerializer = nlohmann::adl_serializer; nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer>::parser_callback_t = std::function<bool(int, nlohmann::detail::parser<nlohmann::basic_json<> >::parse_event_t, nlohmann::basic_json<>&)>]’, declared with attribute ‘nodiscard’ [-Wunused-result]
   48 |         this->meta.parse(s.getText(3));
```

**Comment by pierotofy** [comment]
DSM fixed.

The other errors are due to something else:

```
 96%] Building CXX object CMakeFiles/ddb.dir/vendor/url/Url.cc.o
[ 96%] Linking CXX shared library Release/libddb.so
[ 96%] Built target ddb
Scanning dependencies of target node-ddb
[ 97%] Building CXX object CMakeFiles/node-ddb.dir/nodejs/NativeExtension.cc.o
[ 98%] Building CXX object CMakeFiles/node-ddb.dir/nodejs/ne_dbops.cc.o
/home/runner/work/DroneDB/DroneDB/nodejs/ne_dbops.cc: In member function ‘virtual void RemoveWorker::Execute()’:
/home/runner/work/DroneDB/DroneDB/nodejs/ne_dbops.cc:159:95: error: too many arguments to function ‘DDBErr DDBRemove(const char*, const char**, int)’
  159 |       if (DDBRemove(ddbPath.c_str(), cPaths.data(), static_cast<int>(cPaths.size()), recursive) != DDBERR_NONE){
      |                                                                                               ^
In file included from /home/runner/work/DroneDB/DroneDB/nodejs/ne_dbops.cc:5:
```

File: nodejs/ne_dbops.cc [hunk 3]
```diff
@@ -62,16 +62,16 @@ NAN_METHOD(init) {
 
 class AddWorker : public Nan::AsyncWorker {
  public:
-  AddWorker(Nan::Callback *callback, const std::string &ddbPath, const std::vector<std::string> &paths, bool recursive)
+  AddWorker(Nan::Callback *callback, const std::string &ddbPath, const std::vector<std::string> &paths)
```
Comment (pierotofy) [inline nodejs/ne_dbops.cc:65]: No, no... don't remove the recursive parameter from `ddb add`.

Comment (HeDo88TH) [reply nodejs/ne_dbops.cc]: My bad 👍

File: dotnet/DDB.Tests/Tests.cs [hunk 4]
```diff
@@ -59,7 +59,9 @@ public void testAddRemove()
             Assert.AreEqual(entries.Count, 2);
 
             DroneDB.Remove("testAdd", Path.Join("testAdd", "file.txt"));
-            Assert.Throws<DDBException>(() => DroneDB.Remove("testAdd", "invalid"));
+            
+            // NOTE: Should we throw an exception when the requested file to remove does not exist?
```
Comment (pierotofy) [inline dotnet/DDB.Tests/Tests.cs:63]: Yes.

If a file is not in the index, and a user attempts to remove it, it should throw an error.

```
[piero:/data/test] $ touch test.txt
[piero:/data/test] $ git add test.txt
[piero:/data/test] $ git rm --cached test.txt
rm 'test.txt'
[piero:/data/test] $ git rm --cached test.txt
fatal: pathspec 'test.txt' did not match any files
[piero:/data/test] 128 $ 
```

A bit like git.

Comment (HeDo88TH) [reply dotnet/DDB.Tests/Tests.cs:63]: Ok I will add this behaviour to the `ddb remove` command

Comment (HeDo88TH) [reply dotnet/DDB.Tests/Tests.cs]: Fixed in 14d6103

**Comment by HeDo88TH** [comment]
Do you have any clue why the same test passes on linux and fails on windows? @pierotofy 
It's the nodejs tests

**Comment by pierotofy** [comment]
Not sure; do the tests pass on your Windows machine?

Make sure the `build` directory is empty/removed, then issue:

```
npm install
npm run test
```

Which should build the bindings and run the tests. Maybe you need to increase the time it takes to complete that test?

**Comment by pierotofy** [comment]
Ah, you also need to call `scripts\setup_windows_env.bat` prior to `npm run test`

**Comment by pierotofy** [comment]
Seems to work OK on my machine: 

```
D       DJI_0018.JPG
    V  should be able to call init(), add() and remove() (734ms)
```

**Comment by pierotofy** [comment]
Try to add `    this.timeout(4000); ` in the test body (check `functions.js` for an example), then edit `dbops.js`.

**Comment by HeDo88TH** [comment]
I suppose we are finally ready to merge this masterpiece @pierotofy

**Comment by pierotofy** [comment]
Masterpiece indeed!
